src/Framework/Blocks/BlockCollection.php

```diff
@@ -79,24 +79,106 @@ public function getBlocks(): array
     /**
      * @unreleased
      *
-     * @param BlockModel[] $blocks
-     * @param string $name
-     * @return BlockModel|null
+     * @param string $nameOrId
+     * @param BlockCollection|null $blockCollection
+     * @param string $return self|parent
+     * @return BlockCollection|null
      */
-    public function findByName(string $name, array $blocks = null)
+    public function findBlock(string $nameOrId, BlockCollection $blockCollection = null, string $return = 'self')
     {
-        if ($blocks === null) {
-            $blocks = $this->blocks;
+        if ($blockCollection === null) {
+            $blockCollection = $this;
         }
 
-        foreach ($blocks as $block) {
-            if ($block->name === $name) {
-                return $block;
-            } else if ($block->innerBlocks) {
-                return $this->findByName($name, $block->innerBlocks->blocks);
+        foreach ($blockCollection->blocks as $block) {
+            if ($block->name === $nameOrId || $block->clientId === $nameOrId) {
+                if ($return === 'self') {
+                    return $block->innerBlocks;
+                } elseif ($return === 'parent') {
+                    return $blockCollection;
+                }
+            } elseif ($block->innerBlocks) {
+                $result = $this->findBlock($nameOrId, $block->innerBlocks, $return);
+                if ($result) {
+                    return $result;
+                }
             }
```

Comment:
I think we need to think through this one a bit more.  The `clientId` property is something that changes often (every time the page loads I believe) so we don't want to use that for searching. The `name` property is not a unique attribute as it can be used across multiple blocks which means what we're actually doing here would be finding the first block with a matching name.

This method also looks like it's trying to compensate for being used internally and publicly which is adding complexity.

I would recommend scaling this back to block name to start and see how that simplifies this.  I would also add a method for finding by block index.

This is something that would also highly benefit from unit testing as your building the logic.

Reply:
On 58e267727c469224d1e133ecc915d5885e3ee8bb, I removed the search by ID and implement the logic of $blockIndex so the block used as a reference will be the one presented for the "$blockIndex" time. For example:
```php
// The code below will add the $block as the last child of the third Section block.
$form->blocks->append('givewp/section', $block, 2);
```